### ISSUE TITLE
Refactor TraceGraph

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/slicer#6]
* [x] No need to reinvent monads. We can use:
       * `State` for the `Int` counter
       * `Writer` to accumulate nodes and edges. It is worth thinking whether it makes sense to use writer. On the one hand we will be using `State` already, which allows to accumulate nodes and edges, on the other hand we want don't really care about things in the state, just those in the writer. I think this could be dealt with using a wrapper to run the monad.
       * `State` to store current graph colour. This would eliminate the need to pass around the node arguments
  * [x] It would be good to have a consistent way of dealing with graphing `Trace`, `Exp` and `Syntax`. Type classes might help.
  * [x] I claim:
    ```
    trace2gvG node edge t = traces2gvG (node, node, node) edge t t
    ```
    In other words, a lot of logic is duplicated.
